Add pkg unit tests and mock dependencies

saschagrunert opened the pull request:
This commit adds mocking capabilities via gomock and includes unit tests
for the `pkg` package of cri-o.

TomSweeneyRedHat commented:
/ok-to-test

TomSweeneyRedHat commented on pkg/seccomp/seccomp_test.go:
```diff
@@ -0,0 +1,211 @@
+// +build seccomp
```
cut/paste  and/or leftover debug cruft?

saschagrunert replied:
No this change was intended since the test should only run when seccomp is available on the system.

saschagrunert commented:
/test all

saschagrunert commented:
/test kata-containers

rhatdan commented:
/retest

saschagrunert commented:
/test kata-containers

saschagrunert commented:
/test all

vrothberg commented:
@saschagrunert, the kata tests are flaky, so we can ignore them (for now).

saschagrunert commented:
Okay cool, I just rebased to the latest master to see if everything works as intended.
So please PTAL again.

giuseppe commented:
/test e2e_features_fedora

saschagrunert commented:
/test all

saschagrunert commented:
/test e2e_fedora

saschagrunert commented:
/retest

mrunalp commented on Makefile:
```diff
@@ -171,6 +178,27 @@ testunit: ${BUILD_BIN_PATH}/ginkgo
+	done
+
+mockgen: ${BUILD_BIN_PATH}/mockgen
+	${BUILD_BIN_PATH}/mockgen \
```
Should we use go:generate comments instead of calling mockgen directly? (We can certainly revisit it.)

saschagrunert replied:
Yes we could, that would mean that the build targets looks like:

```make
mockgen: ${BUILD_BIN_PATH}/mockgen
	FLAGS=${MOCKGEN_FLAGS} go generate  ./...
```

and then we could collect the generation commands in `framework.go`:

```go
package framework

//go:generate ../../build/bin/mockgen ${FLAGS} -package containerstoragemock -destination=../mocks/containerstorage/containerstorage.go github.com/containers/storage Store
//go:generate ../../build/bin/mockgen ${FLAGS} -package criostoragemock -destination=../mocks/criostorage/criostorage.go github.com/kubernetes-sigs/cri-o/pkg/storage ImageServer
```

From a build time perspective it makes no difference.

vrothberg commented on pkg/seccomp/seccomp.go:
```diff
@@ -43,7 +43,7 @@ func LoadProfileFromBytes(body []byte, specgen *generate.Generator) error {
 }
 
-var nativeToSeccomp = map[string]Arch{
+var NativeToSeccomp = map[string]Arch{
```
This needs a comment.  It looks like we're exposing internal data just for the sake of testing and that seems wrong but I don't have much experience in those new tests suits so I may miss the point :)

saschagrunert replied:
This is a great discussion point and there are for sure multiple views on 'where should unit testing begin and end'.

From my point of view testing only the internal and external API is fine for unit tests, that's why I put them into a separate package. This will increase the test stability, because I can now refactor the internal package without breaking the tests. The other way around it means every change, which breaks the unit tests, will be an architectural change and is no more refactoring. I think it will also increase the code quality since the API functions have now to be written for being "well testable" too. (:+1: 

I know that this change is a bit hacky and will break dependencies, but for golang projects the "internal package approach" could be used for cri-o too (later). Do you know any projects which use cri-o as dependency? If yes, I would suggest that I add a public function which returns `nativeToSeccomp` instead of adding a comment.

vrothberg replied:
> From my point of view testing only the internal and external API is fine for unit tests, that's why I put them into a separate package. 

If they are separate packages, how can we test the internal API? Will we have separate tests (and packages) for testing the internal and external API?

saschagrunert replied:
> If they are separate packages, how can we test the internal API? 

With "internal" I mean Public functions which are meant to be used internally only. There is no clear separation, yet. But this could be achieved with internal/pkg. Everything which is just "internal" does not need a dedicated API, there are just functions which can be exchanged and are not unit tested. We could test them as well, but I see the pitfall in writing fragile tests which are highly dependent to each other.

> Will we have separate tests (and packages) for testing the internal and external API?

We could. But in the end If a code path is really hard to test because the function is not public I would go for an architectural change: Making smaller functions with dedicated (public / internal) API parts.

vrothberg replied:
> With "internal" I mean Public functions which are meant to be used internally only.

That sounds a bit scary. Having internal APIs doesn't only protect us from project-external uses but also from potential misuse internally (dependencies pile up, tests break because other project-internal packages started to depend on it etc). Unforntunately, `internal/` would only defend us from external use but I absolutely support the idea using that in general.

If we need to make APIs public just testing, then I feel strongly against doing that. In this case, it's not a big deal but I would be concerned if that's not an exception.

Is it possible to keep them in the same package as before or is there a technical reason to break them into a separate one?

vrothberg replied:
But note that's just my 2 cents and I don't want to be the party pooper :) @mrunalp @runcom WDYT?

vrothberg replied:
> I totally agree and it is a major limitation of go, that the separation between internal and external API is not possible.

It depends where we draw the line and how we define internal/external. Seeing it from a visibility/import perspective, I think go is doing a good job. It also allows avoiding external uses with the `./internal` folder. That's pretty much everything I expect from a language in this context.

My main point against making internal APIs public was not only external usage but also internal ones as it can lead to big hairy balls of inter-dependencies, which can turn usually simple changes (e.g., adding/removing parameters) into something big.

> The question I am asking is: Is cri-o a project which should/could/will be used from other libraries as well? I would say "yeah maybe" and therefore switch to the `internal/` pattern.

I don't think it's worth the effort now. Many building blocks of CRI-O are the libraries from `github.com/containers` which have to deal with this kind of problem. But soon, CRI-O will use more parts of `libpod` which entails some (bigger?) changes to the code. I suggest that we revisit after those changes.

> I think this has not been considered for cri-o, yet. For example directly modifying the storage could lead into undefined states and should be only accessible from a well defined public API. So it might be possible that other projects already rely on that.

Are you referring to a concrete API of CRI-O? All-things-storage are usually done in `github.com/containers/storage` which is taking care of consistency, synchronzing storage accesses etc.

> I changed the implementation for the test and made the public var private again.

Thanks!

> I see two open points from our discussion:
> 
> * Should I put the tests into the same package as the implementations?

IMHO yes. I wouldn't like to expose internal APIs for testing (see comments above).

> * Should I open a follow up for restructuring project parts into `internal/`?

I don't think it's worth the effort now as porting CRI-O to use more of `libpod` will entail some bigger changes. It would be sad to invest time into moving around code that is likely to be changed/deleted etc. soon.

Thanks for the nice discussion and the awesome changes to the test infrastructure :+1:

saschagrunert replied:
> I don't think it's worth the effort now. Many building blocks of CRI-O are the libraries from `github.com/containers` which have to deal with this kind of problem. But soon, CRI-O will use more parts of `libpod` which entails some (bigger?) changes to the code. I suggest that we revisit after those changes.

Sounds good :+1: 

> Are you referring to a concrete API of CRI-O? All-things-storage are usually done in `github.com/containers/storage` which is taking care of consistency, synchronizing storage accesses etc.

Okay, a more concrete example from the current master branch: Let's assume I'm implementing a fancy command line tool, which uses cri-o as direct dependency. Now imagine I want to add a new sandbox. I call `AddSandbox()` on the `ContainerServer` I've setup initially from the `lib` package. Then I add it to the pod ID index of the server with `server.PodIDIndex().Add(mySandbox.ID())`.

 Everything fine so far, but I forgot to call `SetInfraContainer()` on that sandbox. If I now call `RemovePodSandbox` from the `server` package it will result in a segfault because of:

https://github.com/kubernetes-sigs/cri-o/blob/e8c2eef9aa7c1bc790d55208da5a201fc7bd6af9/server/sandbox_remove.go#L42

https://github.com/kubernetes-sigs/cri-o/blob/e8c2eef9aa7c1bc790d55208da5a201fc7bd6af9/server/sandbox_remove.go#L65-L69

Oh, shouldn't I call `RunPodSandbox` (`server` package) instead of `AddSandbox()` (`lib` package)? :astonished: Yeah true, but what hampers me in using the stuff from the `lib` package... 🙃

> > * Should I put the tests into the same package as the implementations?
> 
> IMHO yes. I wouldn't like to expose internal APIs for testing (see comments above).

Alright, It should be possible to go for a mixed case here and would not be harmful for the tests. If not possible or needed we can put the tests in the same package as the implementation.

> Thanks for the nice discussion and the awesome changes to the test infrastructure

I have to thank you for taking your time to read through all these words. :+1:

runcom replied:
I've to still look at this PR as a whole but my first reaction to words like "expose internal stuff for testing" made me cry. I'm not saying it's wrong by default, again, I'll have a look at the PR asap, but if we're really doing that there may be some flaw in our code where everything isn't correctly wired in and decoupled to have tests rely just on their own code.
I'll look at this and re-comment in a while!

saschagrunert commented:
During my latest rebase I had to add two new test cases for the latest `GetImageService()` modifications:
https://github.com/kubernetes-sigs/cri-o/blob/980de65a10ff5546b639b4845cd5e57844b7905f/pkg/storage/image_test.go#L87-L95
https://github.com/kubernetes-sigs/cri-o/blob/980de65a10ff5546b639b4845cd5e57844b7905f/pkg/storage/image_test.go#L116-L129

Please have a look again

vrothberg commented:
> During my latest rebase I had to add two new test cases for the latest `GetImageService()` modifications:
> [...]
> Please have a look again

Awesome, thanks! Will have a look now.

vrothberg commented:
LGTM, @runcom @mrunalp, PTAL

runcom commented:
I'm having a look at this today! @saschagrunert sorry if it took so long but great work, much appreciated!

saschagrunert commented:
/test all

saschagrunert commented:
/test e2e_features_rhel

saschagrunert commented:
@runcom I don't want to put the pressure to high so see this as a friendly ping :)

runcom commented:
Yes! No worry for the ping, just got home from conference (on the train home actually lol), this will be the week we'll merge this no worry :)

saschagrunert commented:
/test all

saschagrunert commented:
/test e2e_rhel

saschagrunert commented:
@runcom I rebased on top of the latest master including the vendoring changes. I also added the whitelist for gomock. In the end it seems that some packages get removed nevertheless (see diff).

saschagrunert commented:
/retest

saschagrunert commented:
> @saschagrunert I intend to make a final pass and get this in tomorrow.

That would be really awesome, thank you very much!

mrunalp commented:
/lgtm

mrunalp commented:
/approve
